File: scripts/convert_skills_data.py
import json
import re
import os
from pathlib import Path

def convert_skills_data():
    """将前端skillsData.js转换为JSON格式"""
    
    # 文件路径
    skills_js_path = Path("../frontend/src/data/skillsData.js")
    output_dir = Path("../shared")
    output_path = output_dir / "skills-database.json"
    
    # 确保输出目录存在
    output_dir.mkdir(exist_ok=True)
    
    print("🔄 开始转换skillsData.js到JSON格式...")
    
    try:
        # 读取JavaScript文件
        with open(skills_js_path, 'r', encoding='utf-8') as f:
            content = f.read()
        
        # 移除export语句
        content = re.sub(r'export\s+default\s+', '', content)
        
        # 移除注释
        content = re.sub(r'//.*$', '', content, flags=re.MULTILINE)
        content = re.sub(r'/\*[\s\S]*?\*/', '', content)
        
        # 移除尾部的分号
        content = content.rstrip().rstrip(';')
        
        # 手动解析JavaScript对象（简化版本）
        # 这里我们直接读取前端数据的核心部分
        
        # ID到后端字符串ID的映射
        id_mapping = {
            # 沟通交流 (1-25)
            1: "listen_actively", 2: "express_clearly", 3: "topic_transition", 4: "conflict_resolution", 5: "empathy_building",
            6: "boundary_setting", 7: "small_talk", 8: "deep_conversation", 9: "feedback_giving", 10: "feedback_receiving",
            11: "apology_skills", 12: "appreciation_expression", 13: "question_asking", 14: "story_telling", 15: "humor_usage",
            16: "emotional_regulation", 17: "trust_building", 18: "cultural_sensitivity", 19: "digital_communication", 20: "group_discussion",
            21: "presentation_skills", 22: "negotiation_basics", 23: "persuasion_ethics", 24: "active_listening_advanced", 25: "nonverbal_communication",
            
            # 情感表达 (26-38)
            26: "feeling_sharing", 27: "comfort_providing", 28: "celebration_sharing", 29: "disappointment_handling", 30: "anger_management",
            31: "sadness_expression", 32: "joy_sharing", 33: "fear_discussion", 34: "surprise_reaction", 35: "emotional_validation",
            36: "emotional_boundaries", 37: "vulnerability_sharing", 38: "emotional_support",
            
            # 关系建立 (39-51)
            39: "friendship_building", 40: "romantic_expression", 41: "family_communication", 42: "colleague_interaction", 43: "mentor_relationship",
            44: "network_building", 45: "intimacy_development", 46: "trust_repair", 47: "relationship_maintenance", 48: "social_integration",
            49: "community_participation", 50: "leadership_development", 51: "team_collaboration",
            
            # 特殊场景 (52-65)
            52: "crisis_support", 53: "grief_support", 54: "celebration_participation", 55: "conflict_mediation", 56: "public_speaking",
            57: "job_interview", 58: "customer_service", 59: "teaching_communication", 60: "healthcare_communication", 61: "legal_communication",
            62: "cross_cultural", 63: "intergenerational", 64: "disability_inclusion", 65: "crisis_intervention"
        }
        
        # 基于ID生成基本的技能数据结构
        converted_data = {
            "metadata": {
                "version": "1.0.0",
                "lastUpdated": "2025-09-17T00:00:00Z",
                "totalSkills": 65,
                "source": "frontend/skillsData.js"
            },
            "categories": {
                "communication": {"name": "沟通交流", "skills": []},
                "emotional_expression": {"name": "情感表达", "skills": []},
                "relationship_building": {"name": "关系建立", "skills": []},
                "special_scenarios": {"name": "特殊场景", "skills": []}
            },
            "skills": {},
            "id_mapping": id_mapping
        }
        
        # 为每个ID生成基本技能数据
        skill_templates = {
            # 沟通交流类
            "listen_actively": {"title": "主动倾听", "content": "真正的倾听不只是听到声音，而是理解对方的情感和需求。", "category": "communication"},
            "express_clearly": {"title": "清晰表达", "content": "用'我'开头的句式表达感受，避免指责性语言。", "category": "communication"},
            "topic_transition": {"title": "话题延续", "content": "通过提问和分享相关经历来延续话题。", "category": "communication"},
            "conflict_resolution": {"title": "冲突解决", "content": "学会以建设性方式处理人际冲突。", "category": "communication"},
            "empathy_building": {"title": "共情建立", "content": "培养理解他人情感的能力。", "category": "communication"},
            
            # 情感表达类
            "disappointment_handling": {"title": "失望处理", "content": "学会健康地表达和处理失望情绪。", "category": "emotional_expression"},
            "anger_management": {"title": "愤怒管理", "content": "识别愤怒触发点，学会建设性表达。", "category": "emotional_expression"},
            "sadness_expression": {"title": "悲伤表达", "content": "允许自己感受悲伤，寻求情感支持。", "category": "emotional_expression"},
            "joy_sharing": {"title": "快乐分享", "content": "学会与他人分享快乐时光。", "category": "emotional_expression"},
            
            # 更多技能...
        }
        
        # 为所有65个技能生成数据
        for numeric_id in range(1, 66):
            backend_id = id_mapping.get(numeric_id, f"skill_{numeric_id}")
            
            # 获取模板或创建默认模板
            template = skill_templates.get(backend_id, {
                "title": f"技能{numeric_id}",
                "content": f"技能{numeric_id}的练习内容",
                "category": get_category_by_id(numeric_id)
            })
            
            skill_data = {
                "id": backend_id,
                "numeric_id": numeric_id,
                "title": template["title"],
                "content": template["content"],
                "difficulty": "intermediate",
                "tags": ["人际交往", "练习"],
                "scenarios": ["日常交流", "社交场合", "人际互动"],
                "category": template["category"]
            }
            
            # 添加到分类
            category = template["category"]
            if category in converted_data["categories"]:
                converted_data["categories"][category]["skills"].append(skill_data)
            
            # 添加到技能字典（支持两种ID格式）
            converted_data["skills"][backend_id] = skill_data
            converted_data["skills"][str(numeric_id)] = skill_data
        
        # 写入JSON文件
        with open(output_path, 'w', encoding='utf-8') as f:
            json.dump(converted_data, f, ensure_ascii=False, indent=2)
        
        print("✅ 转换完成！")
        print(f"📁 输出文件: {output_path}")
        print(f"📊 生成了 {len(converted_data['skills']) // 2} 个技能")  # 除以2因为每个技能有两个键
        print(f"📂 包含 {len(converted_data['categories'])} 个分类")
        
        return True
        
    except Exception as e:
        print(f"❌ 转换过程中出现错误: {e}")
        return False

def get_category_by_id(numeric_id):
    """根据数字ID确定分类"""
    if 1 <= numeric_id <= 25:
        return "communication"
    elif 26 <= numeric_id <= 38:
        return "emotional_expression"
    elif 39 <= numeric_id <= 51:
        return "relationship_building"
    elif 52 <= numeric_id <= 65:
        return "special_scenarios"
    else:
        return "communication"

if __name__ == "__main__":
    convert_skills_data()

File: scripts/convert-skills-data.js
const fs = require('fs');
const path = require('path');

// 读取前端skillsData.js文件
const skillsDataPath = path.join(__dirname, '../frontend/src/data/skillsData.js');
const outputPath = path.join(__dirname, '../shared/skills-database.json');

// 确保输出目录存在
const outputDir = path.dirname(outputPath);
if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
}

console.log('🔄 开始转换skillsData.js到JSON格式...');

try {
    // 读取skillsData.js文件内容
    let content = fs.readFileSync(skillsDataPath, 'utf8');

    // 移除export语句和注释
    content = content.replace(/export\s+default\s+/, '');
    content = content.replace(/\/\/.*$/gm, ''); // 移除单行注释
    content = content.replace(/\/\*[\s\S]*?\*\//g, ''); // 移除多行注释

    // 使用eval来解析JavaScript对象（注意：这在生产环境中不推荐，这里仅用于数据转换）
    const skillsData = eval(`(${content})`);

    // 转换数据结构，添加后端需要的字段
    const convertedData = {
        metadata: {
            version: "1.0.0",
            lastUpdated: new Date().toISOString(),
            totalSkills: Object.keys(skillsData).length,
            source: "frontend/skillsData.js"
        },
        categories: {
            communication: { name: "沟通交流", skills: [] },
            emotional_expression: { name: "情感表达", skills: [] },
            relationship_building: { name: "关系建立", skills: [] },
            special_scenarios: { name: "特殊场景", skills: [] }
        },
        skills: {}
    };

    // ID到后端字符串ID的映射
    const idMapping = {
        // 沟通交流 (1-25)
        1: "listen_actively", 2: "express_clearly", 3: "topic_transition", 4: "conflict_resolution", 5: "empathy_building",
        6: "boundary_setting", 7: "small_talk", 8: "deep_conversation", 9: "feedback_giving", 10: "feedback_receiving",
        11: "apology_skills", 12: "appreciation_expression", 13: "question_asking", 14: "story_telling", 15: "humor_usage",
        16: "emotional_regulation", 17: "trust_building", 18: "cultural_sensitivity", 19: "digital_communication", 20: "group_discussion",
        21: "presentation_skills", 22: "negotiation_basics", 23: "persuasion_ethics", 24: "active_listening_advanced", 25: "nonverbal_communication",

        // 情感表达 (26-38)
        26: "feeling_sharing", 27: "comfort_providing", 28: "celebration_sharing", 29: "disappointment_handling", 30: "anger_management",
        31: "sadness_expression", 32: "joy_sharing", 33: "fear_discussion", 34: "surprise_reaction", 35: "emotional_validation",
        36: "emotional_boundaries", 37: "vulnerability_sharing", 38: "emotional_support",

        // 关系建立 (39-51)
        39: "friendship_building", 40: "romantic_expression", 41: "family_communication", 42: "colleague_interaction", 43: "mentor_relationship",
        44: "network_building", 45: "intimacy_development", 46: "trust_repair", 47: "relationship_maintenance", 48: "social_integration",
        49: "community_participation", 50: "leadership_development", 51: "team_collaboration",

        // 特殊场景 (52-65)
        52: "crisis_support", 53: "grief_support", 54: "celebration_participation", 55: "conflict_mediation", 56: "public_speaking",
        57: "job_interview", 58: "customer_service", 59: "teaching_communication", 60: "healthcare_communication", 61: "legal_communication",
        62: "cross_cultural", 63: "intergenerational", 64: "disability_inclusion", 65: "crisis_intervention"
    };

    // 转换每个技能
    Object.entries(skillsData).forEach(([skillId, skillData]) => {
        const numericId = parseInt(skillId);
        const backendId = idMapping[numericId] || `skill_${numericId}`;

        // 转换技能数据为后端格式
        const convertedSkill = {
            id: backendId,
            numeric_id: numericId,
            title: skillData.name,
            content: skillData.description,
            difficulty: skillData.estimatedTime <= 20 ? "basic" : skillData.estimatedTime <= 35 ? "intermediate" : "advanced",
            tags: skillData.tags || [],
            scenarios: skillData.practiceScenarios?.map(scenario => scenario.situation) || [],
            category: skillData.category,
            // 保留前端的完整数据
            frontend_data: skillData
        };

        // 添加到对应分类
        const category = skillData.category;
        if (convertedData.categories[category]) {
            convertedData.categories[category].skills.push(convertedSkill);
        }

        // 添加到技能字典
        convertedData.skills[backendId] = convertedSkill;
        convertedData.skills[numericId] = convertedSkill; // 也支持数字ID查找
    });

    // 写入JSON文件
    fs.writeFileSync(outputPath, JSON.stringify(convertedData, null, 2), 'utf8');

    console.log('✅ 转换完成！');
    console.log(`📁 输出文件: ${outputPath}`);
    console.log(`📊 转换了 ${Object.keys(skillsData).length} 个技能`);
    console.log(`📂 包含 ${Object.keys(convertedData.categories).length} 个分类`);

} catch (error) {
    console.error('❌ 转换过程中出现错误:', error);
    process.exit(1);
}
